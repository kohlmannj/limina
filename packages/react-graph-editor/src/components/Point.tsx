/** @jsx jsx */
import { css, jsx, SerializedStyles } from '@emotion/core';
import { FunctionComponent } from 'react';
import styled, { StyledComponent } from '@emotion/styled';
import { ViewBoxContext } from './ViewBoxContext';
import { ThemeProps, ViewBox, OriginDirection, PointShape } from '..';

export interface PointProps {
  className?: string;
  selected?: boolean;
  x: number;
  y: number;
}

export interface StyledButtonProps extends PointProps, ViewBox {}

const getOriginTranslateFactor = (origin: OriginDirection) =>
  origin === 'left' || origin === 'top' ? -0.5 : 0.5;

export type DynamicStyle = (props: PointProps) => SerializedStyles | undefined;

export const shapes: Record<PointShape, DynamicStyle> = {
  circle: () => css`
    border-radius: 50%;
  `,
  diamond: () => undefined,
  square: () => undefined,
};

const selectedClassName = (props: PointProps & { theme?: ThemeProps }) => css`
  border-color: ${props.theme && props.theme.color ? props.theme.color : '#000'};
  background: #fff;
  box-shadow: 0 0 5px 2px rgba(59, 153, 252, 0.75);
`;

const translate = (props: StyledButtonProps) =>
  props.originX && props.originY
    ? `translate(${getOriginTranslateFactor(props.originX) * 100}%, ${getOriginTranslateFactor(
        props.originY
      ) * 100}%)`
    : '';

const position = (props: StyledButtonProps & { theme?: ThemeProps }) => css`
  position: absolute;
  ${props.originX && `${props.originX}: ${((props.x - props.minX) / props.width) * 100}%;`}
  ${props.originY && `${props.originY}: ${((props.y - props.minY) / props.height) * 100}%;`}
  transform: ${translate(props)} ${
  props.theme && props.theme.shape === 'diamond' ? 'rotate(45deg)' : ''
};
`;

// eslint-disable-next-line @typescript-eslint/no-unused-vars
const Button: FunctionComponent<StyledButtonProps> = ({ className, selected, x, y, ...rest }) => (
  <button
    type="button"
    className={className}
    aria-label={`Point (${x}, ${y})`}
    title={`(${x}, ${y})`}
    {...rest}
  />
);

const StyledButton: StyledComponent<StyledButtonProps, StyledButtonProps, ThemeProps> = styled(
  Button
)`
  width: ${props => `${props.theme.thickness || 8}px`};
  height: ${props => `${props.theme.thickness || 8}px`};
  box-sizing: border-box;
  padding: 0;
  border: 1px solid transparent;
  background: ${props => props.theme.color || '#000'};
  box-shadow: 0 0 1px 1px rgba(0, 0, 0, 0.15);
  cursor: move;
  font-size: 0;
  line-height: 0;
  ${props => (!props.theme || props.theme.shape === 'circle') && `border-radius: 50%`};
  ${props => props.selected && selectedClassName};
  ${position};
  outline: 0;
  text-indent: 100%;
`;

export const Point: FunctionComponent<PointProps> = props => (
  <ViewBoxContext.Consumer>
    {(viewBox: ViewBox) => <StyledButton {...props} {...viewBox} />}
  </ViewBoxContext.Consumer>
);

Point.defaultProps = {
  selected: false,
};
