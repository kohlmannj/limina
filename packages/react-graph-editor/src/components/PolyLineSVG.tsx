import React, { FunctionComponent } from 'react';
import styled, { StyledComponent } from '@emotion/styled';
import { getDimensionsForPoints } from '../utils';
import { ThemeProps, PointTuple } from '..';

export type InterpolationKeyword = 'linear' | 'ease-in-out' | 'ease-in' | 'ease-out';

export type InterpolationFunction = (x: number) => number;

export type InterpolationKeywordOrFunction = InterpolationKeyword | InterpolationFunction;

export interface PolyLineSVGProps {
  className?: string;
  interpolation?: InterpolationKeywordOrFunction;
  label?: string;
  points: PointTuple[];
}

export const PolyLineSVG: FunctionComponent<PolyLineSVGProps> = ({
  className,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  label,
  points,
  ...rest
}) => (
  <svg xmlns="http://www.w3.org/2000/svg" {...getDimensionsForPoints(points)}>
    <polyline
      className={className}
      vectorEffect="non-scaling-stroke"
      points={points.map(point => point.join(',')).join(' ')}
      {...rest}
    />
  </svg>
);

export const StyledPolyLineSVG: StyledComponent<
  PolyLineSVGProps,
  PolyLineSVGProps,
  ThemeProps
> = styled(PolyLineSVG)`
  fill: none;
  stroke: ${props => props.theme.color || '#000'};
  stroke-width: ${props => `${props.theme.thickness / 4 || 2}px`};
`;
