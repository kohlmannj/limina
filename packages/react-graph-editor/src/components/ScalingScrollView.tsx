import React, { Component } from 'react';
import { css } from 'react-emotion';
import { ScrollBarOrientation } from '..';
import ScrollView, { IScrollViewProps } from './ScrollView';

export type ScalePosition = 'start' | 'end';

export interface IScalingScrollGraphState {
  mouseStartX?: number;
  mouseStartY?: number;
  orientation?: ScrollBarOrientation;
  position?: ScalePosition;
  scaleX?: number;
  scaleY?: number;
  thumbLength?: number;
  trackLength?: number;
}

export interface IScaleEventHandlerOptions {
  orientation: ScrollBarOrientation;
  position: ScalePosition;
}

export type MouseEventListener = (event: MouseEvent) => void;

export interface IOnStopScaleEventHandlerFactory extends IScaleEventHandlerOptions {
  onMouseMove: MouseEventListener;
}

export type ScaleEventHandlerFactory = (
  options: IScaleEventHandlerOptions
) => React.MouseEventHandler<HTMLButtonElement>;

export default class StatefulScrollGraph extends Component<
  IScrollViewProps,
  IScalingScrollGraphState
> {
  public static defaultProps = {
    scaleX: 1,
    scaleY: 1,
  };

  public state: IScalingScrollGraphState = {
    scaleX: this.props.scaleX,
    scaleY: this.props.scaleY,
  };

  public componentWillReceiveProps({ scaleX, scaleY }: IScrollViewProps) {
    let shouldStopDragging = false;

    if (scaleX !== this.props.scaleX && typeof scaleX !== 'undefined') {
      this.setState({ scaleX });
      shouldStopDragging = true;
    }

    if (scaleY !== this.props.scaleY && typeof scaleY !== 'undefined') {
      this.setState({ scaleY });
      shouldStopDragging = true;
    }

    if (shouldStopDragging) {
      window.removeEventListener('mousemove', this.onChangeScale);
    }
  }

  public componentWillUnmount() {
    window.removeEventListener('mousemove', this.onChangeScale);
    window.removeEventListener('mouseup', this.onStopScale);
  }

  public render() {
    const {
      children,
      /* scaleX: scaleXProp, scaleY: scaleYProp, */ style /* , ...rest */,
    } = this.props;
    const { scaleX, scaleY } = this.state;

    return (
      <ScrollView
        contentClassName={css`
          width: ${(scaleX || 1) * 100}%;
          height: ${(scaleY || 1) * 100}%;
        `}
        scaleX={scaleX}
        scaleY={scaleY}
        thumbXStartProps={{
          onMouseDown: this.onStartScale({ orientation: 'horizontal', position: 'start' }),
        }}
        thumbXEndProps={{
          onMouseDown: this.onStartScale({ orientation: 'horizontal', position: 'end' }),
        }}
        thumbYStartProps={{
          onMouseDown: this.onStartScale({ orientation: 'vertical', position: 'start' }),
        }}
        thumbYEndProps={{
          onMouseDown: this.onStartScale({ orientation: 'vertical', position: 'end' }),
        }}
        className={css`
          width: 100%;
          height: 100%;
        `}
        style={style}
      >
        {children}
      </ScrollView>
    );
  }

  private onStartScale = ({ orientation, position }: IScaleEventHandlerOptions) => ({
    clientX: mouseStartX,
    clientY: mouseStartY,
    target,
  }: React.MouseEvent<HTMLButtonElement>): void => {
    const thumb: HTMLElement | null = (target as HTMLButtonElement).parentElement;
    const track: HTMLElement | null = thumb ? (thumb as HTMLDivElement).parentElement : null;

    let thumbLength;
    let trackLength;

    if (thumb) {
      const { width, height } = thumb.getBoundingClientRect();

      if (orientation === 'horizontal') {
        thumbLength = width;
      } else if (orientation === 'vertical') {
        thumbLength = height;
      }
    }

    if (track) {
      const { width, height } = track.getBoundingClientRect();

      if (orientation === 'horizontal') {
        trackLength = width;
      } else if (orientation === 'vertical') {
        trackLength = height;
      }
    }

    this.setState({ orientation, mouseStartX, mouseStartY, position, thumbLength, trackLength });

    window.addEventListener('mousemove', this.onChangeScale);
    window.addEventListener('mouseup', this.onStopScale, {
      once: true,
    });
  };

  private onChangeScale = (e: MouseEvent): void => {
    const {
      mouseStartX,
      mouseStartY,
      orientation,
      position,
      thumbLength,
      trackLength,
    } = this.state;

    const dX = e.clientX - mouseStartX!;
    const dY = e.clientY - mouseStartY!;

    let dLength;

    if (orientation === 'horizontal') {
      dLength = position === 'start' ? -dX : dX;
    } else if (orientation === 'vertical') {
      dLength = position === 'start' ? -dY : dY;
    }

    const nextThumbLength = thumbLength && dLength && thumbLength + dLength;
    const rawNextScale = trackLength && nextThumbLength ? trackLength / nextThumbLength : undefined;
    const nextScale = Math.min(Math.max(rawNextScale || 1, 1), 10);

    this.setState({ [`scale${orientation === 'horizontal' ? 'X' : 'Y'}`]: nextScale });
  };

  private onStopScale = (e: MouseEvent): void => {
    this.setState({
      orientation: undefined,
      position: undefined,
      thumbLength: undefined,
      trackLength: undefined,
    });

    window.removeEventListener('mousemove', this.onChangeScale);
  };
}
