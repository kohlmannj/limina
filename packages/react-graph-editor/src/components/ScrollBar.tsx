import React, { CSSProperties, SFC } from 'react';
import styled, { css, StyledComponent } from 'react-emotion';
import { OverflowMode, ScrollBarOrientation } from '..';

export interface IScrollBarProps {
  className?: string;
  orientation?: ScrollBarOrientation;
  overflow?: OverflowMode;
  progress?: number;
  scale?: number;
  style?: CSSProperties;
}

export interface IScrollBarTheme {
  thumbColor?: string;
  thumbWidth?: number;
  trackColor?: string;
  trackWidth?: number;
}

const getOvershootFactor = (progress = 0, min = 0, max = 1): number => {
  let overshoot;

  if (progress < min) {
    overshoot = 1 - Math.abs(progress);
  } else if (progress >= max) {
    overshoot = 1 - Math.abs(progress - max);
  } else {
    overshoot = 1;
  }

  return overshoot;
};

const positionForOrientation = (props: IStyledScrollBarProps) => css`
  top: ${!props.orientation || props.orientation === 'vertical' ? 0 : 'auto'};
  right: ${!props.orientation || props.orientation === 'vertical'
    ? 0
    : `${
        props.theme && typeof props.theme.trackWidth === 'number' ? props.theme.trackWidth : 15
      }px`};
  bottom: ${!props.orientation || props.orientation === 'vertical'
    ? `${props.theme && typeof props.theme.trackWidth === 'number' ? props.theme.trackWidth : 15}px`
    : 0};
  left: ${!props.orientation || props.orientation === 'vertical' ? 'auto' : 0};
`;

const Track = styled<IScrollBarProps, 'div'>('div')`
  display: flex;
  flex-direction: ${props =>
    !props.orientation || props.orientation === 'vertical' ? 'column' : 'row'};
  align-items: center;
  justify-content: ${props =>
    typeof props.progress === 'number' && props.progress >= 0.5 ? 'flex-end' : 'flex-start'};
  background: ${props => props.theme.trackColor || `#efefef`};
  ${props =>
    !props.orientation || props.orientation === 'vertical' ? 'width' : 'height'}: ${props =>
    `${typeof props.theme.trackWidth === 'number' ? props.theme.trackWidth : 15}px`};
  ${positionForOrientation};
`;

const ThumbContainer = styled<IScrollBarProps, 'div'>('div')`
  display: flex;
  flex-direction: inherit;
  align-items: center;
  justify-content: inherit;
  ${props =>
    !props.orientation || props.orientation === 'vertical' ? 'width: 100%' : 'height: 100%'};
  ${props =>
    !props.orientation || props.orientation === 'vertical' ? 'height' : 'width'}: ${props =>
    `calc(1 / ${props.scale || 1} * ${getOvershootFactor(props.progress)} * 100%)`};
  transform: ${props =>
    `translate${!props.orientation || props.orientation === 'vertical' ? 'Y' : 'X'}(${(Math.min(
      Math.max(props.progress || 0, 0),
      1
    ) +
      (typeof props.progress === 'number' && props.progress >= 0.5 ? -1 : 0)) *
      ((props.scale || 1) - 1) *
      100}%)`};
`;

export interface IStyledScrollBarProps extends IScrollBarProps {
  theme?: IScrollBarTheme;
}

export const Thumb = styled<IScrollBarProps, 'button'>('button')`
  display: flex;
  flex: 0 1 100%;
  flex-direction: inherit;
  padding: 0;
  border: 0;
  margin: ${props =>
    `${!props.orientation || props.orientation === 'vertical' ? '' : '0 '}${((typeof props.theme
      .trackWidth === 'number'
      ? props.theme.trackWidth
      : 15) -
      (typeof props.theme.thumbWidth === 'number' ? props.theme.thumbWidth : 15)) /
      2}px${!props.orientation || props.orientation === 'vertical' ? ' 0' : ''}`};
  background: ${props => props.theme.thumbColor || '#666'};
  border-radius: ${props =>
    `${(typeof props.theme.thumbWidth === 'number' ? props.theme.thumbWidth : 15) / 2}px`};
  ${props =>
    !props.orientation || props.orientation === 'vertical' ? 'width' : 'height'}: ${props =>
    `${typeof props.theme.thumbWidth === 'number' ? props.theme.thumbWidth : 15}px`};
`;

export const ScrollBar: SFC<IScrollBarProps> = ({
  className,
  orientation,
  overflow,
  progress,
  scale,
  style,
}) => (
  <Track className={className} orientation={orientation} progress={progress} style={style}>
    <ThumbContainer orientation={orientation} progress={progress} scale={scale}>
      {typeof scale === 'number' &&
        (scale > 1 || (scale <= 1 && overflow === 'scroll')) && <Thumb orientation={orientation} />}
    </ThumbContainer>
  </Track>
);

ScrollBar.defaultProps = {
  orientation: 'vertical',
  overflow: 'auto',
  scale: 1,
};

export default ScrollBar;
