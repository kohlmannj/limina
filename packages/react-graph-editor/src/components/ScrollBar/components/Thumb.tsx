import React from 'react';
import styled, { StyledComponent } from 'react-emotion';
import { IScrollBarProps } from '..';
import { defaultTheme } from '../../../theme';

const Thumb = styled<IScrollBarProps, 'button'>('button')`
  display: flex;
  flex: 0 1 100%;
  flex-direction: inherit;
  padding: 0;
  border: 0;
  margin: ${props =>
    `${!props.orientation || props.orientation === 'vertical' ? '' : '0 '}${((typeof props.theme
      .trackWidth === 'number'
      ? props.theme.trackWidth
      : 15) -
      (typeof props.theme.thumbWidth === 'number'
        ? props.theme.thumbWidth
        : defaultTheme.thumbWidth)) /
      2}px${!props.orientation || props.orientation === 'vertical' ? ' 0' : ''}`};
  background: ${props => props.theme.thumbColor || defaultTheme.thumbColor};
  border-radius: ${props =>
    `${(typeof props.theme.thumbWidth === 'number'
      ? props.theme.thumbWidth
      : defaultTheme.thumbWidth) / 2}px`};
  ${props =>
    !props.orientation || props.orientation === 'vertical' ? 'width' : 'height'}: ${props =>
    `${
      typeof props.theme.thumbWidth === 'number' ? props.theme.thumbWidth : defaultTheme.thumbWidth
    }px`};
`;

export default Thumb;
