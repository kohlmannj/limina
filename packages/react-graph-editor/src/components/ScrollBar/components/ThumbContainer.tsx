/** @jsx jsx */
import { css, jsx } from '@emotion/core';
import { FunctionComponent } from 'react';
import { ScrollBarProps, StyledScrollBarProps } from '..';
import { defaults } from '../../../scrollBarTheme';

export const getOvershootFactor = (progress = 0, min = 0, max = 1): number => {
  let overshoot;

  if (progress < min) {
    overshoot = 1 - Math.abs(progress);
  } else if (progress > max) {
    overshoot = 1 - Math.abs(progress - max);
  } else {
    overshoot = 1;
  }

  return overshoot;
};

export const getTranslateFactor = ({ progress, scale }: ScrollBarProps) => {
  const cappedProgress = Math.min(Math.max(progress || 0, 0), 1);
  const halfwayAddend = typeof progress !== 'undefined' && progress > 0.5 ? -1 : 0;
  const scaleMultiplier = (scale || 1) - 1;
  return (cappedProgress + halfwayAddend) * scaleMultiplier;
};

export const getMainAxisLength = ({ axis }: ScrollBarProps) => ({
  [!axis || axis === 'y' ? 'width' : 'height']: '100%',
});

export const getCrossAxisLength = ({ axis, progress, scale }: ScrollBarProps) => {
  const crossAxisProperty = !axis || axis === 'y' ? 'height' : 'width';
  // const minCrossAxisProperty = `min${crossAxisProperty}`;

  return {
    [crossAxisProperty]: `${(1 / (scale || 1)) * getOvershootFactor(progress) * 100}%`,
  };
};

export const getSingleBorderRadiusValue = ({ theme }: StyledScrollBarProps) =>
  (theme && typeof theme.thumbWidth !== 'undefined' ? theme.thumbWidth : defaults.thumbWidth) / 2;

export const getTransform = ({ axis, progress, scale }: ScrollBarProps) => ({
  transform: `translate${(axis || 'y').toUpperCase()}(${getTranslateFactor({
    progress,
    scale,
  }) * 100}%)`,
});

export const getSingleMarginValue = ({ theme }: StyledScrollBarProps) =>
  ((theme && typeof theme.trackWidth !== 'undefined' ? theme.trackWidth : defaults.trackWidth) -
    (theme && typeof theme.thumbWidth !== 'undefined' ? theme.thumbWidth : defaults.thumbWidth)) /
  2;

const thumbCss = css`
  position: relative;
  display: flex;
  width: 100%;
  height: 100%;
  flex-direction: inherit;
  align-items: stretch;
  justify-content: center;

  &::before {
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    margin: ${getSingleMarginValue({})}px;
    background: ${defaults.thumbColor};
    border-radius: ${getSingleBorderRadiusValue({})}px;
    content: '';
  }
`;

export const ThumbContainer: FunctionComponent<ScrollBarProps> = ({
  axis,
  progress,
  scale,
  style,
  ...rest
}) => (
  <div
    css={thumbCss}
    style={{
      ...style,
      ...getMainAxisLength({ axis }),
      ...getCrossAxisLength({ axis, progress, scale }),
      ...getTransform({ axis, progress, scale }),
    }}
    {...rest}
  />
);
