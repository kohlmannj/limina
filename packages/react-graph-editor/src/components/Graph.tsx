/** @jsx jsx */
import { jsx } from '@emotion/core';
import chroma from 'chroma-js';
import { ThemeProvider } from 'emotion-theming';
import { CSSProperties, Fragment, FunctionComponent } from 'react';
import styled, { StyledComponent } from '@emotion/styled';
import { getViewBoxForPoints } from '../utils';
import { Point, shapes } from './Point';
import { PolyLineSVGProps } from './PolyLineSVG';
import { ViewBoxContext } from './ViewBoxContext';
import { ThemeProps, OriginDirection, PointTuple } from '..';

const shapeNames = Object.keys(shapes);

export interface GraphProps {
  className?: string;
  lines: PolyLineSVGProps[];
  minX?: number;
  minY?: number;
  maxX?: number;
  maxY?: number;
  originX?: OriginDirection;
  originY?: OriginDirection;
  style?: CSSProperties;
}

const Graph: FunctionComponent<GraphProps> = ({
  className,
  lines,
  minX: minXProp,
  minY: minYProp,
  maxX: maxXProp,
  maxY: maxYProp,
  originX,
  originY,
  ...rest
}) => {
  const autoViewBox = getViewBoxForPoints(
    lines.reduce((reducedPoints: PointTuple[], { points }) => [...reducedPoints, ...points], [])
  );

  const minX = typeof minXProp === 'number' ? minXProp : autoViewBox.minX;
  const minY = typeof minYProp === 'number' ? minYProp : autoViewBox.minY;
  const maxX = typeof maxXProp === 'number' ? maxXProp : autoViewBox.maxX;
  const maxY = typeof maxYProp === 'number' ? maxYProp : autoViewBox.maxY;

  const viewBox = {
    minX,
    minY,
    maxX,
    maxY,
    originX,
    originY,
    width: typeof minX === 'number' && typeof maxX === 'number' ? Math.abs(maxX - minX) : 0,
    height: typeof minY === 'number' && typeof maxY === 'number' ? Math.abs(maxY - minY) : 0,
  };

  return (
    <div className={className} {...rest}>
      <ViewBoxContext.Provider value={viewBox}>
        {lines.map(({ label, points }, index) => (
          <ThemeProvider
            key={label}
            theme={{
              color: chroma.brewer.Set1[index % chroma.brewer.Set1.length],
              shape: shapeNames[index % shapeNames.length],
            }}
          >
            <Fragment>
              {points.map(([x, y]) => (
                <Point key={`${label}_${x}_${y}`} x={x} y={y} />
              ))}
            </Fragment>
          </ThemeProvider>
        ))}
      </ViewBoxContext.Provider>
    </div>
  );
};

Graph.defaultProps = {
  originX: 'left',
  originY: 'bottom',
};

const StyledGraph: StyledComponent<GraphProps, GraphProps, ThemeProps> = styled(Graph)`
  position: relative;
  width: 100%;
  height: 100%;
`;

const GraphContainer: FunctionComponent<GraphProps> = ({ className, style, ...rest }) => (
  <div className={className} style={style}>
    <StyledGraph {...rest} />
  </div>
);

export const StyledGraphContainer = styled(GraphContainer)`
  width: 100%;
  height: 100%;
  box-sizing: border-box;
  padding: ${props => `${(props.theme.thickness || 8) / 2}px`};
`;
