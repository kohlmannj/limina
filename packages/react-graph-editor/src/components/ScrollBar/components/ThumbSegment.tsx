/** @jsx jsx */
import { css, jsx } from '@emotion/core';
import { ButtonHTMLAttributes, FunctionComponent } from 'react';
import { ScrollBarProps, StyledScrollBarProps } from '..';
import { defaults } from '../../../scrollBarTheme';
import { getSingleBorderRadiusValue, getSingleMarginValue } from './ThumbContainer';

export type ThumbSegmentPosition = 'start' | 'middle' | 'end';

export interface ThumbSegmentAdditionalProps {
  dragSignifier?: boolean;
  position: ThumbSegmentPosition;
}

export interface ThumbSegmentProps
  extends ButtonHTMLAttributes<HTMLButtonElement>,
    ThumbSegmentAdditionalProps,
    ScrollBarProps {}

export interface StyledThumbSegmentProps extends ThumbSegmentProps, StyledScrollBarProps {}

export const dragSignifier = (props: StyledThumbSegmentProps) =>
  css({
    cursor: `${!props.axis || props.axis === 'y' ? 'row' : 'col'}-resize`,

    '&::after': {
      position: 'absolute',
      top: 0,
      right: 0,
      bottom: 0,
      left: 0,
      boxSizing: 'border-box',
      margin: 'auto',
      [!props.axis || props.axis === 'y' ? 'borderTop' : 'borderLeft']: '1px solid',
      [!props.axis || props.axis === 'y' ? 'borderBottom' : 'borderRight']: '1px solid',
      borderColor:
        props.theme && props.theme.trackColor ? props.theme.trackColor : defaults.trackColor,
      [!props.axis || props.axis === 'y' ? 'width' : 'height']: `${(props.theme &&
      typeof props.theme.thumbWidth !== 'undefined'
        ? props.theme.thumbWidth
        : defaults.thumbWidth) / 2}px`,
      [!props.axis || props.axis === 'y' ? 'height' : 'width']: `${(props.theme &&
      typeof props.theme.thumbWidth !== 'undefined'
        ? props.theme.thumbWidth
        : defaults.thumbWidth) / 4}px`,
      [!props.axis || props.axis === 'y' ? 'minHeight' : 'minWidth']: '3px',
      content: "''",
    },
  });

export const getBorderRadiusValue = ({ axis, position, theme }: StyledThumbSegmentProps) => {
  const radius = getSingleBorderRadiusValue({ theme });
  let borderRadiusValue = '0';

  if (!axis || axis === 'y') {
    if (position === 'start') {
      borderRadiusValue = `${radius} ${radius} 0 0`;
    } else if (position === 'end') {
      borderRadiusValue = `0 0 ${radius} ${radius}`;
    }
  } else if (position === 'start') {
    borderRadiusValue = `${radius} 0 0 ${radius}`;
  } else if (position === 'end') {
    borderRadiusValue = `0 ${radius} ${radius} 0`;
  }

  return borderRadiusValue;
};

export const getFlex = (props: StyledThumbSegmentProps) => {
  switch (props.position) {
    case 'start':
    case 'end':
      return `0 0 ${(props.theme && typeof props.theme.thumbWidth !== 'undefined'
        ? props.theme.thumbWidth
        : defaults.thumbWidth) +
        getSingleMarginValue(props) * 2}px`;
    case 'middle':
    default:
      return '0 1 100%';
  }
};

const thumbSegment = ({ position, theme }: StyledThumbSegmentProps) => css`
  position: relative;
  flex: ${getFlex({ position, theme })};
  padding: ${getSingleMarginValue({ theme })}px;
  border: 0;
  margin: 0;
  background: transparent;
  border-radius: ${getBorderRadiusValue({ position, theme })}px;
`;

const thumbSegmentClassNames = {
  start: thumbSegment({ position: 'start' }),
  middle: thumbSegment({ position: 'middle' }),
  end: thumbSegment({ position: 'end' }),
};

const dragSignifierClassNames = {
  x: {
    start: dragSignifier({ axis: 'x', position: 'start' }),
    middle: dragSignifier({ axis: 'x', position: 'middle' }),
    end: dragSignifier({ axis: 'x', position: 'end' }),
  },
  y: {
    start: dragSignifier({ axis: 'y', position: 'start' }),
    middle: dragSignifier({ axis: 'y', position: 'middle' }),
    end: dragSignifier({ axis: 'y', position: 'end' }),
  },
};

export const ThumbSegment: FunctionComponent<ThumbSegmentProps> = ({
  dragSignifier: dragSignifierProp,
  axis,
  position,
  ...rest
}) => (
  <button
    type="button"
    css={[
      thumbSegmentClassNames[position],
      dragSignifierProp && dragSignifierClassNames[axis || 'y'][position],
    ]}
    {...rest}
  />
);
