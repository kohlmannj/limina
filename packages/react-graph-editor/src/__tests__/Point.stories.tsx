import { storiesOf } from '@storybook/react';
import { ThemeProvider } from 'emotion-theming';
import React from 'react';
import { Point } from '../components/Point';

storiesOf('react-graph-editor/Point', module)
  .add('with default appearance', () => <Point x={0} y={0} />)
  .add('with selected appearance', () => <Point selected x={0} y={0} />)
  .add('with diamond appearance', () => (
    <ThemeProvider theme={{ shape: 'diamond' }}>
      <Point x={0} y={0} />
    </ThemeProvider>
  ))
  .add('with diamond, selected appearance', () => (
    <ThemeProvider theme={{ shape: 'diamond' }}>
      <Point selected x={0} y={0} />
    </ThemeProvider>
  ))
  .add('with square appearance', () => (
    <ThemeProvider theme={{ shape: 'square' }}>
      <Point x={0} y={0} />
    </ThemeProvider>
  ))
  .add('with square, selected appearance', () => (
    <ThemeProvider theme={{ shape: 'square' }}>
      <Point selected x={0} y={0} />
    </ThemeProvider>
  ))
  .add('with theme', () => (
    <ThemeProvider theme={{ color: 'red', thickness: 16 }}>
      <Point x={0} y={0} />
    </ThemeProvider>
  ))
  .add('with theme and selected appearance', () => (
    <ThemeProvider theme={{ color: 'red', thickness: 16 }}>
      <Point selected x={0} y={0} />
    </ThemeProvider>
  ));
