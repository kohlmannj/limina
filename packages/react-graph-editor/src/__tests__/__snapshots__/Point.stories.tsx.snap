// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react-graph-editor/Point with default appearance 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`react-graph-editor/Point with diamond appearance 1`] = `
<Component
  selected={false}
  x={0}
  y={0}
/>
`;

exports[`react-graph-editor/Point with diamond, selected appearance 1`] = `
<Component
  selected={true}
  x={0}
  y={0}
/>
`;

exports[`react-graph-editor/Point with selected appearance 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`react-graph-editor/Point with square appearance 1`] = `
<Component
  selected={false}
  x={0}
  y={0}
/>
`;

exports[`react-graph-editor/Point with square, selected appearance 1`] = `
<Component
  selected={true}
  x={0}
  y={0}
/>
`;

exports[`react-graph-editor/Point with theme 1`] = `
<Component
  selected={false}
  x={0}
  y={0}
/>
`;

exports[`react-graph-editor/Point with theme and selected appearance 1`] = `
<Component
  selected={true}
  x={0}
  y={0}
/>
`;
